### ISSUE TITLE
[ttx] AttributeError "isCFF2" from T2CharString when dumping CFF2 toXML()

### ISSUE THREAD

**Issue by anthrotype** [fonttools/fonttools#1451]
Dumping "SourceSerifVariable-Roman.otf" with ttx yields an AttributeError from T2CharString.toXML method:

```
$ ttx ~/Downloads/SourceSerifVariable-Roman.otf 
Dumping "/home/clupo/Downloads/SourceSerifVariable-Roman.otf" to "/home/clupo/Downloads/SourceSerifVariable-Roman.ttx"...
Dumping 'GlyphOrder' table...
Dumping 'head' table...
Dumping 'hhea' table...
Dumping 'maxp' table...
Dumping 'OS/2' table...
Dumping 'hmtx' table...
Dumping 'cmap' table...
Dumping 'name' table...
Dumping 'post' table...
Dumping 'BASE' table...
Dumping 'CFF2' table...
ERROR: Unhandled exception has occurred
Traceback (most recent call last):
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttx.py", line 398, in main
    process(jobs, options)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttx.py", line 372, in process
    action(input, output, options)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/loggingTools.py", line 375, in wrapper
    return func(*args, **kwds)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttx.py", line 269, in ttDump
    newlinestr=options.newlinestr)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/ttFont.py", line 228, in saveXML
    self._saveXML(writer, **kwargs)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/ttFont.py", line 283, in _saveXML
    self._tableToXML(tableWriter, tag, splitGlyphs=splitGlyphs)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/ttFont.py", line 314, in _tableToXML
    table.toXML(writer, self)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/tables/C_F_F_.py", line 42, in toXML
    self.cff.toXML(writer)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 156, in toXML
    font.toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 2330, in toXML
    BaseDict.toXML(self, xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 2272, in toXML
    conv.xmlWrite(xmlWriter, name, value)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 1175, in xmlWrite
    value.toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 722, in toXML
    self[i].toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 2272, in toXML
    conv.xmlWrite(xmlWriter, name, value)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 1175, in xmlWrite
    value.toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 2272, in toXML
    conv.xmlWrite(xmlWriter, name, value)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 1175, in xmlWrite
    value.toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 652, in toXML
    subr.toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 1091, in toXML
    if self.isCFF2:
AttributeError: 'T2CharString' object has no attribute 'isCFF2'
```

If we need to parametrize the toXML method depending on whether isCFF2 or not, then we need to have an extra isCFF2=False parameter in the top-level CFFFontSet class, and set that to True from the `table_C_F_F__2.toXML`

**Comment by anthrotype** [comment]
this looks more involved that I had thought.

So, I removed the check in toXML for extra args on stack in #1452, and I can now dump to SourceSerifProVF to ttx.
However, after I compile back to OTF and attempt to re-dump to TTX, I get a failure while de-compiling:

```
$ ttx '/home/clupo/Downloads/SourceSerifVariable-Roman#5.otf' 
Dumping "/home/clupo/Downloads/SourceSerifVariable-Roman#5.otf" to "/home/clupo/Downloads/SourceSerifVariable-Roman#16.ttx"...
Dumping 'GlyphOrder' table...
Dumping 'head' table...
Dumping 'hhea' table...
Dumping 'maxp' table...
Dumping 'OS/2' table...
Dumping 'hmtx' table...
Dumping 'cmap' table...
Dumping 'name' table...
Dumping 'post' table...
Dumping 'BASE' table...
Dumping 'CFF2' table...
ERROR: Error in charstring 237
ERROR: Unhandled exception has occurred
Traceback (most recent call last):
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttx.py", line 398, in main
    process(jobs, options)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttx.py", line 372, in process
    action(input, output, options)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/loggingTools.py", line 375, in wrapper
    return func(*args, **kwds)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttx.py", line 269, in ttDump
    newlinestr=options.newlinestr)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/ttFont.py", line 228, in saveXML
    self._saveXML(writer, **kwargs)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/ttFont.py", line 283, in _saveXML
    self._tableToXML(tableWriter, tag, splitGlyphs=splitGlyphs)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/ttFont.py", line 314, in _tableToXML
    table.toXML(writer, self)
  File "/home/clupo/Github/fonttools/Lib/fontTools/ttLib/tables/C_F_F_.py", line 42, in toXML
    self.cff.toXML(writer)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 156, in toXML
    font.toXML(xmlWriter)
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 2322, in toXML
    self.decompileAllCharStrings()
  File "/home/clupo/Github/fonttools/Lib/fontTools/cffLib/__init__.py", line 2336, in decompileAllCharStrings
    charString.decompile()
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 966, in decompile
    decompiler.execute(self)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 300, in execute
    rv = handler(index)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 338, in op_callsubr
    self.execute(subr)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 292, in execute
    token, isOperator, index = charString.getToken(index)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 1053, in getToken
    token, index = handler(self, b0, self.bytecode, index)
  File "/home/clupo/Github/fonttools/Lib/fontTools/misc/psCharStrings.py", line 33, in read_smallInt1
    b1 = byteord(data[index])
IndexError: index out of range
``` 

I believe I haven't touched the decompile part of the code. I'm not sure what's wrong...

**Comment by justvanrossum** [comment]
Does this problem persist with the latest version of  #1452?

**Comment by anthrotype** [comment]
unfortunately it does...

to reproduce, download "SourceSerifVariable-Roman.otf" then dump to ttx, then compile back to otf, then attempt to re-dump to ttx. You should get the above `IndexError: index out of range` exception at line 33 of psCharStrings.py.

https://github.com/adobe-fonts/source-serif-pro/blob/release/VAR/SourceSerifVariable-Roman.otf

**Comment by justvanrossum** [comment]
I'm on to something, debugging.

**Comment by justvanrossum** [comment]
Found it, making PR.

**Comment by anthrotype** [comment]
Fixed by #1456

**Comment by justvanrossum** [comment]
Closed via 176f2e82449c5bce47e6b63b52ec674582a1e12b
